### PR TITLE
[Counterfactual] Deploy safe with first transaction

### LINKED ISSUE #3184: [Counterfactual] Adjust first transaction flow
## Scope

- Deploy the safe with the first transaction
- Show gas estimation
- Clear counterfactual state after deployment

## Out of scope

- Design changes

### PR DESCRIPTION
**Pull request by usame-algan**
## What it solves

Resolves #3184 

## How this PR fixes it

- Deploys safe with first transaction
- Shows gas estimation for the bundle
- Clear counterfactual state after deployment
- Hide nonce in all transaction flows

## ToDos

- [x] Adjust AppFrame to handle incoming requests for counterfactual safes

## How to test it

1. Create a counterfactual 1/1 safe
2. Send some funds to it
3. Create a transaction
4. Gas estimation should be displayed
5. Execute the transaction
6. Observe you have to sign first
7. Observe the safe is deployed after execution
8. Observe the tx history is visible after execution

## Checklist
* [ ] I've tested the branch on mobile 📱
* [ ] I've documented how it affects the analytics (if at all) 📊
* [ ] I've written a unit/e2e test for it (if applicable) 🧑‍💻

### CODE REVIEW

File: src/features/counterfactual/hooks/useDeployGasLimit.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  const chainId = useChainId()
+
+  const [gasLimit, gasLimitError, gasLimitLoading] = useAsync<bigint | undefined>(async () => {
+    if (!safeTx || !onboard) return
+
+    const sdk = await getSafeSDKWithSigner(onboard, chainId)
+
+    const gas = await estimateTxBaseGas(sdk, safeTx)
+    const safeTxGas = await estimateSafeTxGas(sdk, safeTx)
+    const safeDeploymentGas = await estimateSafeDeploymentGas(sdk)
```
Comment (katspaugh) [inline src/features/counterfactual/hooks/useDeployGasLimit.ts:19]: These things don't depend on each other so they could be done in parallel.

File: src/features/counterfactual/utils.ts [hunk 2]
```diff
@@ -35,3 +41,46 @@ export const getUndeployedSafeInfo = (undeployedSafe: PredictedSafeProps, addres
...
+    const wallet = await assertWalletChain(onboard, chainId)
+
+    const signedTx = await sdkUnchecked.signTransaction(safeTx)
+    const deploymentTx = await sdkUnchecked.wrapSafeTransactionIntoDeploymentBatch(signedTx, txOptions)
+
+    const provider = createWeb3(wallet.provider)
+    const signer = await provider.getSigner()
+
+    // @ts-ignore TODO: Check if the other type also works
+    result = await signer.sendTransaction(deploymentTx)
```
Comment (katspaugh) [inline src/features/counterfactual/utils.ts:64]: Some of these calls could be parallelized as well, not sure how fast it is already.

File: src/features/counterfactual/hooks/useDeployGasLimit.ts (hunk 1 above)
Comment (usame-algan) [reply src/features/counterfactual/hooks/useDeployGasLimit.ts]: Good idea!

File: src/features/counterfactual/utils.ts (hunk 2 above)
Comment (usame-algan) [reply src/features/counterfactual/utils.ts]: We could run the `wallet` and `deploymentTx` call in parallel but I think its already fast currently

**Review by katspaugh — APPROVED** [review]
👍

**Comment by francovenica** [comment]
LGTM

I tried creating a CF safe, transfer funds to it and then sending them.
After the execution, in the history tab, you can see the safe creation tx and the send funds tx as well.

**Pull request merged**